src/rdkafka_admin.c

```diff
@@ -2735,3 +2837,400 @@ rd_kafka_DescribeConfigs_result_resources (
+                             rd_kafka_topic_partition_list_free);
+                rd_list_add(&rko_result->rko_u.admin_result.results,
+                            rd_kafka_topic_partition_list_copy(offsets));
+                *rko_resultp = rko_result;
+                rd_kafka_AdminFanoutState_free(state);
+                return RD_KAFKA_RESP_ERR_NO_ERROR;
+        } else {
+                /* Still waiting for outstanding requests to return */
+                return RD_KAFKA_RESP_ERR__PARTIAL;
+        }
+
+err_parse:
+        if (rko_result)
+                rd_kafka_op_destroy(rko_result);
+
+        rd_snprintf(errstr, errstr_size,
+                    "DeleteRecords response protocol parse failure: %s",
+                    rd_kafka_err2str(err));
+
+        return err;
+}
+
+rd_kafka_resp_err_t rd_kafka_DeleteRecords (rd_kafka_t *rk,
+                                            rd_kafka_topic_partition_list_t *offsets,
+                                            const rd_kafka_AdminOptions_t *options,
+                                            rd_kafka_queue_t *rkqu,
+                                            int timeout_ms) {
+        static const struct rd_kafka_admin_worker_cbs cbs = {
+                rd_kafka_DeleteRecordsRequest,
+                rd_kafka_DeleteRecordsResponse_parse,
+        };
+        rd_kafka_resp_err_t err;
+        int i;
+        rd_list_t leaders;
+        struct rd_kafka_partition_leader *leader;
+
+
+        rd_list_init(&leaders, offsets->cnt,
+                     (void *)rd_kafka_partition_leader_destroy);
+        err = rd_kafka_topic_partition_list_query_leaders(rk, offsets, &leaders,
+                                                          timeout_ms);
+
+        if (err) {
+                rd_list_destroy(&leaders);
+                return err;
+        }
+
+        rd_kafka_AdminFanoutState_t *state =
+                rd_kafka_AdminFanoutState_new(rd_list_cnt(&leaders),
+                                              rd_kafka_topic_partition_list_free,
+                                              rd_kafka_topic_partition_list_free);
+
+        rd_kafka_topic_partition_list_t *results =
+                rd_kafka_topic_partition_list_new(offsets->cnt);
+        rd_list_add(&state->responses, results);
+
+        /* For each leader send a request for its partitions */
+        RD_LIST_FOREACH(leader, &leaders, i) {
+                rd_kafka_op_t *rko;
+
+                rko = rd_kafka_admin_request_op_new(rk,
+                                                    RD_KAFKA_OP_DELETERECORDS,
+                                                    RD_KAFKA_EVENT_DELETERECORDS_RESULT,
+                                                    &cbs, options, rkqu);
+                rko->rko_u.admin_request.broker_id = leader->rkb->rkb_nodeid;
+                rd_kafka_AdminOptions_set_opaque(&rko->rko_u.admin_request.options, state);
```

Comment:
This'll overwrite the application's opaque, suggest adding a new field to the internal AdminOptions for this purpose.
Don't need accessor functions for it since it is just used internally.

Reply:
Add a test-case to verify that the user's opaque makes it through to the result.